WIP: ENT-211 add aggregate course key

zubair-arbi opened the pull request:
@asadiqbal08 @saleem-latif @mattdrayer 
Add a new type `AggregateCourseKey` of `OpaqueKey` for handling course keys without run information.

zubair-arbi commented:
@cpennington Please let me know if this implementation is what you have suggested.

cpennington commented on opaque_keys/edx/locator.py:
```diff
@@ -1330,3 +1330,67 @@ def to_deprecated_list_repr(self):
 AssetKey.set_deprecated_fallback(AssetLocator)
+
+
+class AggregateCourseLocator(AggregateCourseKey):    # pylint: disable=abstract-method
```
What formats of key does this need to parse? Can you add some tests to make sure it parses/serializes the way you expect?

Also, should you be able to get an `AggregateCourseKey` from a `CourseKey`?

zubair-arbi replied:
@cpennington @clintonb  The only format need is `org+course`. I have added/update tests in `opaque_keys/edx/tests/test_locators.py` for the `AggregateCourseLocator`.

I was unable to directly get a `AggregateCourseKey` locator from a `CourseKey` locator mainly due to changed `KEY_TYPE` and [`OpaqueKey` Hierarchy](https://github.com/edx/edx-platform/wiki/Opaque-Keys-(Locators)#opaquekey-hierarchy). To handle the conversion from `CourseKey` to `AggregateCourseKey` I have added a new method `from_course_key` in `AggregateCourseKey` which takes a course key object and returns a locator `AggregateCourseLocator` from it. Please let me know your suggestion on this approach.

zubair-arbi replied:
a) Sorry I forgot to mention that the complete format of course key has version, e.g. `course-v2:edX+DemoX`
b) I have added the method `make_course_key_v2` in `CourseKey` which will return `CourseKeyV2` locator. Also added the method `make_course_run_key` in `CourseKeyV2` which will return a `CourseKey` locator `CourseLocator` from the provided course run.
@cpennington thanks for the suggestions. Please have a look on the changes and let me know if I have missed anything.

zubair-arbi commented on opaque_keys/edx/locator.py:
```diff
@@ -1330,3 +1330,67 @@ def to_deprecated_list_repr(self):
+        """
+        The serialized course key without run information.
+        """
+        return '{org}+{course}'.format(org=self.org, course=self.course)
```
Removed the method `serialize`

clintonb commented on opaque_keys/edx/tests/test_locators.py:
```diff
@@ -59,3 +61,44 @@ def test_version_tree(self):
+        aggregate_course_key = 'aggregate-course:org+course'
+        aggregate_course_locator = AggregateCourseKey.from_string(aggregate_course_key)
+        serialized_aggregate_course_key = 'org+course'
+        # Allow access to _to_string
```
This comment isn't necessary.

clintonb commented on setup.py:
```diff
@@ -2,7 +2,7 @@
 setup(
     name='edx-opaque-keys',
-    version='0.4',
+    version='0.5',
```
0.5.0

clintonb commented on opaque_keys/edx/tests/test_locators.py:
```diff
@@ -59,3 +61,44 @@ def test_version_tree(self):
+    Tests for :class:`.AggregateCourseLocator`
+    """
+
+    def test_aggregate_course_locator(self):
```
Add docstrings explaining what's being tested in each of these methods.

zubair-arbi replied:
added

clintonb commented on opaque_keys/edx/tests/test_locators.py:
```diff
@@ -59,3 +61,44 @@ def test_version_tree(self):
+        'course-v1:org+course+run',
+    )
+    def test_aggregate_course_locator_from_course_key(self, course_id):
+        course = CourseKey.from_string(course_id)
```
s/course/course_key

clintonb commented on opaque_keys/edx/tests/test_locators.py:
```diff
@@ -59,3 +61,44 @@ def test_version_tree(self):
+        aggregate_course_locator = AggregateCourseLocator.from_course_key(course)
+
+        aggregate_course_key = 'aggregate-course:{}+{}'.format(course.org, course.course)
+        self.assertEqual(aggregate_course_locator, AggregateCourseKey.from_string(aggregate_course_key))
```
The use of `AggregateCourseKey.from_string` is confusing and unnecessary.

```python
course_key = CourseKey.from_string(course_id)
expected = AggregateCourseKey(org=course_key.org, course=course_key.course)
actual = AggregateCourseKey.from_course_key(course_key)
assert expected == actual
```

clintonb commented on opaque_keys/edx/tests/test_locators.py:
```diff
@@ -59,3 +61,44 @@ def test_version_tree(self):
+        aggregate_course_key = 'aggregate-course:{}+{}'.format(course.org, course.course)
+        self.assertEqual(aggregate_course_locator, AggregateCourseKey.from_string(aggregate_course_key))
+
+    def test_aggregate_course_locator_serialize(self):
```
`test_serialize_to_string`

clintonb commented on opaque_keys/edx/tests/test_locators.py:
```diff
@@ -59,3 +61,44 @@ def test_version_tree(self):
+
+    def test_aggregate_course_locator_serialize(self):
+        aggregate_course_key = 'aggregate-course:org+course'
+        aggregate_course_locator = AggregateCourseKey.from_string(aggregate_course_key)
```
You should not need to use `AggregateCourseKey.from_string` to instantiate a new instance of `AggregateCourseKey`. Fix this here, and in all other tests where it is used unnecessarily.

zubair-arbi replied:
@clintonb @cpennington This seems to be the [documented way](https://github.com/edx/edx-platform/wiki/Opaque-Keys-(Locators)#parse-course-ids) if you try to initialize the `AggregateCourseKey` directly you get error:
```python
>>> AggregateCourseKey(org=course_key.org, course=course_key.course)
TypeError: Can't instantiate abstract class AggregateCourseKey with abstract
methods _to_string, from_course_key
```

clintonb commented on opaque_keys/edx/locator.py:
```diff
@@ -1330,3 +1330,72 @@ def to_deprecated_list_repr(self):
+            **kwargs
+        )
+
+        if self.org is None or self.course is None:
```
You need to account for empty string: `if not (self.org and self.course)`

clintonb commented on opaque_keys/edx/locator.py:
```diff
@@ -1330,3 +1330,72 @@ def to_deprecated_list_repr(self):
+        """
+        Return a string representing this location.
+        """
+        return u'{org}+{course}'.format(
```
No need for the `u`.

cpennington commented on opaque_keys/edx/keys.py:
```diff
@@ -77,6 +77,26 @@ def block_type(self):  # pragma: no cover
+    An :class:`opaque_keys.OpaqueKey` identifying a
+    serialized Course Key object.
+    """
+    KEY_TYPE = 'aggregate_course_key'
```
No technical issues that I can think of.

mattdrayer replied:
Is this something we should do as part of this PR?  We can certainly make the change but I don't want to introduce two new ideas at one time if we don't have to

clintonb replied:
I would call this course-v2 now. Add a ticket to your backlog to add course-run-v1.

zubair-arbi replied:
@cpennington @clintonb  I have to change the key type to `course_key_v2` or just `course_v2` because the `CANONICAL_NAMESPACE` of `AggregateCourseLocator` would be `course-v2`? Also should I change the name of aggregate course key and locator names too?

zubair-arbi replied:
I have change the `CANONICAL_NAMESPACE` of `AggregateCourseLocator` to `course-v2`. Now the aggregate course key format would be `course-v2:org+course`
```python
AggregateCourseKey.from_string('course-v2:edX+DemoX')
```

clintonb replied:
I thought we agreed to get rid of the word aggregate. I have no clue what it means in this context.

zubair-arbi commented on opaque_keys/edx/locator.py:
```diff
@@ -1330,3 +1330,72 @@ def to_deprecated_list_repr(self):
+
+        Arguments:
+            org (string): Organization identifier for the course
+            course (string): Course name
```
My understanding is we call the middle part of key as course or course name `organization+course_name+course_run` and all together are considered as a unique identifier for a course. I can change it if its confusing.

zubair-arbi replied:
I have changed it to `Course number` as per `Studio`.

zubair-arbi commented:
@cpennington @clintonb please give another review.

clintonb commented on opaque_keys/edx/tests/test_locators.py:
```diff
@@ -59,3 +61,59 @@ def test_version_tree(self):
+
+    def test_aggregate_course_locator(self):
+        """
+        Verify that the method "from_string" of class "AggregateCourseKey"
```
Use ` instead of " for Python code (classes, methods).

zubair-arbi commented:
@clintonb @cpennington I have updated the naming. Please review.

clintonb commented on opaque_keys/edx/keys.py:
```diff
@@ -77,6 +77,26 @@ def block_type(self):  # pragma: no cover
+    __slots__ = ()
+
+    @abstractproperty
+    def from_course_key(self, course_key):  # pragma: no cover
```
Call this `from_course_run_key`. Eventually, course-key-v1 will be replaced with course-run-key-v1.

clintonb commented on opaque_keys/edx/tests/test_locators.py:
```diff
@@ -59,3 +61,59 @@ def test_version_tree(self):
+
+    def test_course_locator_v2(self):
+        """
+        Verify that the method "from_string" of class "CourseKeyV2"
```
Use ` instead of " to format Python entities. This should be fixed in all of your docstrings in this file.

clintonb commented on opaque_keys/edx/tests/test_locators.py:
```diff
@@ -59,3 +61,59 @@ def test_version_tree(self):
+        self.assertEqual(expected_course_locator, course_locator_v2)
+
+    @ddt.data(
+        'org/course/run',
```
There is no need for `ddt` in this instance.  Whether we use the old or new format for `CourseKey` doesn't matter. The method simply expects an instance of `CourseKey`.

clintonb commented on opaque_keys/edx/tests/test_locators.py:
```diff
@@ -59,3 +61,59 @@ def test_version_tree(self):
+        course_key = CourseKey.from_string('course-v1:org+course+run')
+        course_locator_v2 = CourseLocatorV2(org=course_key.org, course=course_key.course)
+        expected_serialized_key = '{org}+{course}'.format(org=course_key.org, course=course_key.course)
+        # pylint: disable=protected-access
```
No need to call the protected member. Use `str(course_locator_v2)`.

clintonb commented on opaque_keys/edx/tests/test_locators.py:
```diff
@@ -59,3 +61,59 @@ def test_version_tree(self):
+        Verify that the method "_to_string" of class "CourseLocatorV2"
+        serializes a course key v2 to a string with expected format.
+        """
+        course_key = CourseKey.from_string('course-v1:org+course+run')
```
This is unnecessary.

clintonb commented on opaque_keys/edx/tests/test_locators.py:
```diff
@@ -59,3 +61,59 @@ def test_version_tree(self):
+    Tests for :class:`.CourseLocatorV2`
+    """
+
+    def test_course_locator_v2(self):
```
`test_from_string`

clintonb commented on opaque_keys/edx/tests/test_locators.py:
```diff
@@ -59,3 +61,59 @@ def test_version_tree(self):
+        'org/course/run',
+        'course-v1:org+course+run',
+    )
+    def test_course_locator_v2_from_course_key(self, course_id):
```
You don't need to prefix your test methods with `course_locator_v2`. The fact that `CourseLocatorV2` is given by the name of the test class, `CourseLocatorV2Tests`. This should simply be `test_from_course_key`.

clintonb commented on opaque_keys/edx/tests/test_locators.py:
```diff
@@ -59,3 +61,59 @@ def test_version_tree(self):
+        'org+course+run+foo',
+        'course-v2:org+course+run',
+    )
+    def test_invalid_format_course_key(self, course_key):
```
`test_from_string_with_invalid_input`

zubair-arbi commented:
@clintonb Please have another look.
@cpennington I have added inter-conversion methods for course key and course key v2. I have updated the sample usage in the description of this PR. Since the `LibraryLocator` is course agnostic so the method `make_course_key_v2` will just raise error `NotImplementedError`.

clintonb commented on opaque_keys/edx/keys.py:
```diff
@@ -61,6 +61,13 @@ def make_asset_key(self, asset_type, path):  # pragma: no cover
         raise NotImplementedError()
 
+    @abstractmethod
+    def make_course_key_v2(self):  # pragma: no cover
```
Is this really necessary given that you already have `CourseKeyV2.from_course_run_key()`?

zubair-arbi replied:
@cpennington @clintonb just to confirm, I have to remove this method only or do I need to remove the `make_course_run_key` in class `CourseKeyV2` too?

cpennington replied:
@zubair-arbi: You should have some a method to go from `CourseKeyV2` to `CourseKey` and back again, but only one of each. I would suggest using regular methods, rather than class methods, in general. So, probably `CourseKeyV2.make_course_run_key`, and `CourseKey.make_course_key_v2`.

clintonb commented on opaque_keys/edx/keys.py:
```diff
@@ -77,6 +84,37 @@ def block_type(self):  # pragma: no cover
+        raise NotImplementedError()
+
+    @abstractproperty
+    def make_course_run_key(self, course_run):  # pragma: no cover
```
Why is this needed? We already have existing methods of creating course run keys.

cpennington replied:
These were my suggestion. The existing keys typically are related by giving the more general key a way to create more specific keys by hydrating additional information.

cpennington replied:
That said, I agree that it doesn't make much sense to have both methods. We should pick one, and stick to it.

cpennington commented:
I'm concerned about the names `CourseKeyV2` and `CourseKey`. They are fundamentally different things, but that distinction isn't really captured in the names. I think it would be awesome to deprecate `CourseKey` and start encouraging the use of `CourseRunKey` and `AggregateCourseKey` (which are clearly different). "Course" is just too ambiguous right now, until a vast majority of our software switches to using CourseRun for course runs.
